{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "received-helen",
   "metadata": {},
   "source": [
    "# Identifying Differentially Expressed Genes for MixScape Pipeline\n",
    "\n",
    "I've personally found that the normal t-test method MixScape doesn't work that detect a large number of differentially expressed genes in comparison to using DESeq2 via psuedo bulking. As such I am using DESeq2 via psuedo bulking to identify differentially expressed genes\n",
    "\n",
    "## Loading Package Dependencies\n",
    "\n",
    "Following section of code setups the computational environment to conduct Quality Control and Merge experimental replicates. First I load the R packages that my analysis is dependent upon. Next are the utility functions to conduct repeative/routine steps of anaylsis. Finally, I load the merged SCE Objects and gRNA Count matrices."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cognitive-creek",
   "metadata": {
    "name": "setup",
    "scrolled": true,
    "tags": [
     "remove_cell"
    ]
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.3.1 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.3.6     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.7     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.9\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.2.0     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 2.1.2     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.1\n",
      "\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    transpose\n",
      "\n",
      "\n",
      "NOTE: Either Arial Narrow or Roboto Condensed fonts are required to use these themes.\n",
      "\n",
      "      Please use hrbrthemes::import_roboto_condensed() to install Roboto Condensed and\n",
      "\n",
      "      if Arial Narrow is not on your system, please see https://bit.ly/arialnarrow\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:rlist’:\n",
      "\n",
      "    List\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘AnnotationDbi’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Data Cleaning packages\n",
    "library(rlist)\n",
    "library(tidyverse)\n",
    "library(data.table)\n",
    "library(hrbrthemes)\n",
    "library(RColorBrewer)\n",
    "library(org.Hs.eg.db)\n",
    "\n",
    "# Single Cell Experiment Packages\n",
    "library(DT)\n",
    "library(fgsea)\n",
    "library(scran)\n",
    "library(scater)\n",
    "library(DESeq2)\n",
    "library(Seurat)\n",
    "library(ggvenn)\n",
    "library(viridis)\n",
    "library(pheatmap)\n",
    "library(gridExtra)\n",
    "library(clusterProfiler)\n",
    "library(SingleCellExperiment)\n",
    "\n",
    "data_dir = '../data'\n",
    "padj_palette = brewer.pal(n = 7, name = \"Reds\")\n",
    "logfc_palette = brewer.pal(n = 7, name = \"RdBu\")\n",
    "\n",
    "genes_levels <- c('MYC', 'MTOR', 'NCOA3', 'TFAP2C', 'GATA3', 'FOXA1', 'MED1', 'ESR1',\n",
    "                  'CCND1', 'SMC1A', 'CREBBP', 'AKT1', 'MED12', 'TP53', 'CARM1',\n",
    "                  'SMARCA4', 'NCOA5', 'EP300', 'PTEN', 'RBI', 'E2F1', 'GDF2', 'NPSR1',\n",
    "                  'BPIFB3', 'CLDN17', 'FCRL4', 'IRGC', 'MEP1A', 'PRB1', 'SPATA16', 'PNPLA5')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "extraordinary-mexico",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_metadata = read.csv(paste0(data_dir, '/Metadata/gene_metadata.csv'))\n",
    "colnames(gene_metadata) = c('Gene_Symbol', 'Category', 'Essentiality')\n",
    "gene_metadata$Gene_Symbol = as.character(gene_metadata$Gene_Symbol)\n",
    "rownames(gene_metadata) = gene_metadata$Gene_Symbol\n",
    "\n",
    "all_genes = gene_metadata %>% pull(Gene_Symbol)\n",
    "expressed_genes = gene_metadata %>% filter(Essentiality != 'Non-essential') %>% pull(Gene_Symbol)\n",
    "er_early = read.csv(paste0(data_dir, '/Metadata/er_early_response.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "covered-plain",
   "metadata": {},
   "source": [
    "## Differential Expression Analysis - Utility Functions \n",
    "\n",
    "The function below implents a series of utility of functions to conduct and analysis differentially expressed analysis. \n",
    "\n",
    "- **conduct_dea_with_deseq2:** Conduct initial modelling of gene expression using DESeq2 and filter the initial matrix by removing all genes with less 10. Finally, aftering modeling gene expression per conduction we save the DESeq2 object.\n",
    "\n",
    "- **fetch_dea_results:** With the provided DESeq2 object extract the DEA results using the results function and then shrink the log fold change using the lcfShrink function both are from the DESeq package. To do this I loop through all predefined conductions (in this case target genes). Once all of the results are extracted I then save the extracted DEA results.\n",
    "\n",
    "- **find_deg_of_target_genes:** This function is responsible for extracting all unique differentially expressed gene from all target genes. What i mean is that if a gene is differentially expressed for any target gene I want to extract that. I then return this list. The extracted list will be used to extract genes in the extract_de_results and generate a heatmap.\n",
    "\n",
    "- **extract_de_results:** Using the list returned from find_deg_of_target_genes we extract log fold change and adjusted p-values to generate a heatmap. adjusted p-values are also log transformed in order to better visual differences in expression.\n",
    "\n",
    "- **heatmap_matrix:** Finally, using the values from differentially expressed genes generate a heatmap showing the differences in gene expression between the different target genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "american-fellowship",
   "metadata": {},
   "outputs": [],
   "source": [
    "conduct_dea_with_deseq2 = function(dds, cache_fp) {\n",
    "  # Remove genes with low counts\n",
    "  keep = rowSums(counts(dds)) >= 10\n",
    "  dds = dds[keep,]\n",
    "  \n",
    "  # estimate size factors and train deseq2 model\n",
    "  dds = estimateSizeFactors(dds)\n",
    "  dds = DESeq(dds)\n",
    "  \n",
    "  # save and return the DEA results\n",
    "  saveRDS(dds, cache_fp)\n",
    "  return (dds)\n",
    "}\n",
    "\n",
    "fetch_dea_results = function(dds, dea_res_fp, gene_col='gene', control_label='Control'){\n",
    "  # fetching the results for all shRNA and applying shrinkage to there log2 foldchange estimates.\n",
    "  \n",
    "  result = list()\n",
    "  \n",
    "  # fetching dea results for all target genes\n",
    "  for(gene in unique(dds[[gene_col]])){\n",
    "    print (gene)\n",
    "    # skip control\n",
    "    if (gene == control_label){\n",
    "      next()\n",
    "    }\n",
    "    \n",
    "    gene_contrast = c(gene_col, gene, control_label)\n",
    "    gene_coefficient = paste0(gene_col, \"_\", gene, \"_vs_\", control_label)\n",
    "      \n",
    "    print(paste0(\"Fetching the Results for \", gene))\n",
    "    res_unshrunken = results(dds, alpha = 0.05, contrast = gene_contrast)\n",
    "    result[[gene]] = lfcShrink(dds, gene_coefficient, res=res_unshrunken, type=\"normal\")\n",
    "  }\n",
    "  \n",
    "  # save and returned fetched dea results\n",
    "  saveRDS(result, dea_res_fp)\n",
    "  return (result)\n",
    "}\n",
    "\n",
    "find_deg_of_target_genes = function(dea_results, expr_genes, col_of_interest='padj', threshold=0.05) {\n",
    "  # looping through all target genes and extracting statistically significant differentially expressed genes. Return the union of all gene names. \n",
    "  deg_gene_list = c()\n",
    "  \n",
    "  for (target_gene in names(dea_results)) {\n",
    "    \n",
    "    if(!target_gene %in% expr_genes){\n",
    "        message(paste0(\"Non-Expressed Gene:\", target_gene, \"...\"))\n",
    "        next()\n",
    "    }\n",
    "    \n",
    "    res = dea_results[[target_gene]]\n",
    "    gene_degs = res[!is.na(res[[col_of_interest]]) & res[[col_of_interest]] < threshold, ]\n",
    "    deg_gene_list = c(deg_gene_list, rownames(gene_degs))\n",
    "  }\n",
    "  \n",
    "  return(deg_gene_list)\n",
    "}\n",
    "\n",
    "extract_deg_results = function(deg_results, list_of_degs, gene_metadata, padj_col='padj', logfc_col='log2FoldChange'){\n",
    "  # Extracting Differential Expression Results of Target Genes\n",
    "  df = data.frame()\n",
    "  deg_target_genes = names(deg_results)\n",
    "  \n",
    "  for(target_gene in rownames(gene_metadata)) {\n",
    "    \n",
    "    if (!target_gene %in% deg_target_genes){\n",
    "      next()\n",
    "    }\n",
    "    \n",
    "    res = as.data.frame(deg_results[[target_gene]][list_of_degs, ])\n",
    "    res[[padj_col]][is.na(res[[padj_col]])] = 1\n",
    "    \n",
    "    target_df = data.frame(\n",
    "      logfc = res[[logfc_col]],\n",
    "      log_padj = -log(res[[padj_col]], 10),\n",
    "      target_gene = rep(target_gene, length(list_of_degs)),\n",
    "      deg = list_of_degs\n",
    "    )\n",
    "    \n",
    "    df = rbind(df, target_df)\n",
    "  }\n",
    "  \n",
    "  df$logfc[is.na(df$logfc)] = 0 \n",
    "  \n",
    "  return (df)\n",
    "}\n",
    "\n",
    "heatmap_matrix = function (df, degs, col_of_interest, transpose=TRUE) {\n",
    "    mat <- list()\n",
    "    \n",
    "    for (gene in unique(df$target_gene)) {\n",
    "        gene_df <- df %>% filter(target_gene == gene)\n",
    "        \n",
    "        vec_df <- gene_df[, col_of_interest] %>% as.data.frame\n",
    "        colnames(vec_df) <- gene\n",
    "        # print (vec_df)\n",
    "        # print (length(gene_df$deg))\n",
    "        rownames(vec_df) <- gene_df$deg\n",
    "        mat[[gene]] <- vec_df\n",
    "    }\n",
    "    \n",
    "    mat = list.cbind(mat) %>% as.matrix\n",
    "    \n",
    "    if (transpose == TRUE) {\n",
    "        mat = t(mat)\n",
    "    }\n",
    "\n",
    "    return (mat)\n",
    "}\n",
    "\n",
    "cache_degs = function(results, library, fp_dir){\n",
    "    # differential expression results for mixscape\n",
    "    degs = list()\n",
    "    de_adj_fp = paste0(fp_dir, '/DEGs/', library, '_degs_mixscape.Rds')\n",
    "\n",
    "    for (target_gene in names(results)){\n",
    "        gene_df = results[[target_gene]]\n",
    "\n",
    "        genes_to_keep = !is.na(gene_df$padj) & gene_df$padj <= 0.05\n",
    "        deg = rownames(gene_df[genes_to_keep, ])\n",
    "\n",
    "        degs[[target_gene]] = deg\n",
    "    }\n",
    "    \n",
    "    saveRDS(degs, de_adj_fp)\n",
    "    return (degs)\n",
    "}\n",
    "\n",
    "# utility functions to count degs in dea results and conduct gene set enrichment analysis\n",
    "deg_count = function (results) {\n",
    "    gene_names = c()\n",
    "    deg_counts = c()\n",
    "\n",
    "    for (gene in names(results)){\n",
    "        res = results[[gene]]\n",
    "        res$padj[is.na(res$padj)] = 1\n",
    "        qc_count = nrow(res[res$padj <= 0.05, ])\n",
    "\n",
    "        gene_names = c(gene_names, gene)\n",
    "        deg_counts = c(deg_counts, qc_count)\n",
    "    }\n",
    "\n",
    "    gene_deg_count = data.frame(deg_count = deg_counts, genes=gene_names)\n",
    "    gene_deg_order = order(gene_deg_count$deg_count, decreasing=TRUE)\n",
    "    \n",
    "    gene_deg_count = gene_deg_count[gene_deg_order, ]\n",
    "    gene_deg_count$genes = as.character(gene_deg_count$genes)\n",
    "    gene_deg_count$genes = factor(gene_deg_count$genes, level=gene_deg_count$genes)\n",
    "    \n",
    "    return (gene_deg_count)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "educated-junction",
   "metadata": {},
   "source": [
    "# Conducting Differential Expression on the Library Level\n",
    "\n",
    "## Submitting a Given Libraries SCE for Differential Expression Analysis on Cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "serious-greene",
   "metadata": {},
   "outputs": [],
   "source": [
    "yuza_fp = paste0(data_dir, '/DEGs/results/Yuza.Rds')\n",
    "merged_fp = paste0(data_dir, '/DEGs/results/Merged.Rds')\n",
    "edwards_fp = paste0(data_dir, '/DEGs/results/Edwards.Rds')\n",
    "brunello_fp = paste0(data_dir, '/DEGs/results/Brunello.Rds')\n",
    "\n",
    "if (!file.exists(yuza_fp)) {\n",
    "    system(\"sbatch ../scripts/using_slurm_for_dea.R Yuza\")\n",
    "} else {\n",
    "    yuza_res = readRDS(yuza_fp)\n",
    "}\n",
    "\n",
    "if (!file.exists(merged_fp)) {\n",
    "   system(\"sbatch ../scripts/using_slurm_for_dea.R Merged\")\n",
    "} else {\n",
    "    merged_res = readRDS(merged_fp)\n",
    "}\n",
    "\n",
    "if (!file.exists(edwards_fp)) {\n",
    "    system(\"sbatch ../scripts/using_slurm_for_dea.R Edwards\")\n",
    "} else {\n",
    "    edwards_res = readRDS(edwards_fp)\n",
    "}\n",
    "\n",
    "if (!file.exists(brunello_fp)) {\n",
    "    system(\"sbatch ../scripts/using_slurm_for_dea.R Brunello\")\n",
    "} else {\n",
    "    brunello_res = readRDS(brunello_fp)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "royal-division",
   "metadata": {},
   "outputs": [],
   "source": [
    "yuza_filtered = cache_degs(yuza_res, 'Yuza', data_dir)\n",
    "merged_filtered = cache_degs(merged_res, 'Merged', data_dir)\n",
    "edwards_filtered = cache_degs(edwards_res, 'Edwards', data_dir)\n",
    "brunello_filtered = cache_degs(brunello_res, 'Brunello', data_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "exterior-things",
   "metadata": {},
   "outputs": [],
   "source": [
    "yuza_degs = find_deg_of_target_genes(yuza_res, all_genes)\n",
    "edwards_degs = find_deg_of_target_genes(edwards_res, all_genes)\n",
    "brunello_degs = find_deg_of_target_genes(brunello_res, all_genes)\n",
    "merged_degs = find_deg_of_target_genes(merged_res, all_genes)\n",
    "\n",
    "# creating a universial set of degs for the analysis\n",
    "uni_degs = unique(c(yuza_degs, edwards_degs, brunello_degs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "transparent-radical",
   "metadata": {},
   "outputs": [],
   "source": [
    "pval_col = 'padj'\n",
    "lfc = 'log2FoldChange'\n",
    "yuza_df = extract_deg_results(yuza_res, uni_degs, gene_metadata, padj_col=pval_col, logfc_col=lfc)\n",
    "edwards_df = extract_deg_results(edwards_res, uni_degs, gene_metadata, padj_col=pval_col, logfc_col=lfc)\n",
    "brunello_df = extract_deg_results(brunello_res, uni_degs, gene_metadata, padj_col=pval_col, logfc_col=lfc)\n",
    "merged_df = extract_deg_results(merged_res, merged_degs, gene_metadata, padj_col=pval_col, logfc_col=lfc)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "quick-command",
   "metadata": {},
   "source": [
    "### DEG Count Barplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "entertaining-colleague",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_genes = merged_df %>% pull(target_gene) %>% unique() %>% as.character\n",
    "yuza_genes = yuza_df %>% pull(target_gene) %>% unique() %>% as.character\n",
    "edwards_genes = edwards_df %>% pull(target_gene) %>% unique() %>% as.character\n",
    "brunello_genes = brunello_df %>% pull(target_gene) %>% unique() %>% as.character\n",
    "\n",
    "merged_deg_count = deg_count(merged_res)\n",
    "yuza_deg_count = deg_count(yuza_res)\n",
    "edwards_deg_count = deg_count(edwards_res)\n",
    "brunello_deg_count = deg_count(brunello_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "about-fault",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_deg_plot = ggplot(merged_deg_count, aes(x=genes, y=deg_count)) +\n",
    "                  geom_col(stat=\"identity\", position=position_dodge()) +\n",
    "                  coord_flip() + \n",
    "                  theme_bw() +\n",
    "                  scale_x_discrete(limits = rev(levels(merged_deg_count$genes))) +\n",
    "                  theme(plot.background = element_blank(),\n",
    "                      panel.grid.major = element_blank(),\n",
    "                      panel.grid.minor = element_blank(),\n",
    "                      panel.border = element_blank(),\n",
    "                      axis.line = element_line(color = 'black')) +\n",
    "                  labs(x='Target Genes', y='Number of Detected DEGs',\n",
    "                       title='Number of DEGs Detected with All Libraries Merged')\n",
    "\n",
    "yuza_deg_plot = ggplot(yuza_deg_count, aes(x=reorder(genes, deg_count), y=deg_count)) +\n",
    "                  geom_col(stat=\"identity\", position=position_dodge()) +\n",
    "                  ylim(0, 450) +\n",
    "                  coord_flip() + \n",
    "                  theme_bw() +\n",
    "                  scale_x_discrete(limits = rev(levels(merged_deg_count$genes))) +\n",
    "                  theme(plot.background = element_blank(),\n",
    "                      panel.grid.major = element_blank(),\n",
    "                      panel.grid.minor = element_blank(),\n",
    "                      panel.border = element_blank(),\n",
    "                      axis.line = element_line(color = 'black')) +\n",
    "                  labs(x='Target Genes', y='Number of Detected DEGs',\n",
    "                       title='Number of DEGs Detected for the Yuza Library')\n",
    "\n",
    "edwards_deg_plot = ggplot(edwards_deg_count, aes(x=reorder(genes, deg_count), y=deg_count)) +\n",
    "                  geom_col(stat=\"identity\", position=position_dodge()) +\n",
    "                  ylim(0, 450) +\n",
    "                  coord_flip() + \n",
    "                  theme_bw() +\n",
    "                  scale_x_discrete(limits = rev(levels(merged_deg_count$genes))) +\n",
    "                  theme(plot.background = element_blank(),\n",
    "                      panel.grid.major = element_blank(),\n",
    "                      panel.grid.minor = element_blank(),\n",
    "                      panel.border = element_blank(),\n",
    "                      axis.line = element_line(color = 'black')) +\n",
    "                  labs(x='Target Genes', y='Number of Detected DEGs',\n",
    "                       title='Number of DEGs Detected for the Edwards Library')\n",
    "\n",
    "brunello_deg_plot = ggplot(brunello_deg_count, aes(x=reorder(genes, deg_count), y=deg_count)) +\n",
    "                      geom_col(stat=\"identity\", position=position_dodge()) +\n",
    "                      ylim(0, 450) + \n",
    "                      coord_flip() + \n",
    "                      theme_bw() +\n",
    "                      scale_x_discrete(limits = rev(levels(merged_deg_count$genes))) +\n",
    "                      theme(plot.background = element_blank(),\n",
    "                          panel.grid.major = element_blank(),\n",
    "                          panel.grid.minor = element_blank(),\n",
    "                          panel.border = element_blank(),\n",
    "                          axis.line = element_line(color = 'black')) +\n",
    "                      labs(x='Target Genes', y='Number of Detected DEGs',\n",
    "                           title='Detected DEG Count for the Brunello Library')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acoustic-photographer",
   "metadata": {},
   "outputs": [],
   "source": [
    "yuza_deg_count$library <- 'Yuza'\n",
    "edwards_deg_count$library <- 'Edwards'\n",
    "brunello_deg_count$library <- 'Brunello'\n",
    "\n",
    "genes_levels <- c('TFAP2C','MYC','GATA3','MTOR','CCND1','SMC1A','ESR1',\n",
    "                  'FOXA1','NCOA3','CREBBP','MED1','MED12','AKT1','SMARCA4',\n",
    "                  'EP300','CARM1','TP53','NCOA5', 'RB1', 'E2F1', 'ACCSL','MEP1A',\n",
    "                  'IRGC','NPSR1','PNPLA5','BPIFB3','GDF2','SPATA16','CLDN17','FCRL4')\n",
    "\n",
    "deg_count <- bind_rows(yuza_deg_count, edwards_deg_count, brunello_deg_count)\n",
    "deg_count$genes <- factor(deg_count$genes, levels=genes_levels)\n",
    "deg_count$library <- factor(deg_count$library, levels=c('Edwards', 'Brunello', 'Yuza'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ordered-nation",
   "metadata": {},
   "outputs": [],
   "source": [
    "deg_count_plot <- deg_count %>% \n",
    "                        ggplot(aes(x=genes, y=deg_count, fill=library)) +\n",
    "                              geom_col(stat=\"identity\", position=position_dodge()) +\n",
    "                              ylim(0, 450) + \n",
    "                              coord_flip() + \n",
    "                              theme_bw() +\n",
    "                              scale_fill_manual(values=rev(c('#529EFF', '#F8766D', '#5BB300'))) +\n",
    "                              scale_x_discrete(limits = rev(levels(deg_count$genes))) +\n",
    "                              theme(panel.border = element_blank(),\n",
    "                                    plot.background = element_blank(),\n",
    "                                    panel.grid.major = element_blank(),\n",
    "                                    panel.grid.minor = element_blank(),\n",
    "                                    axis.line = element_line(color = 'black')) +\n",
    "                              labs(x='Target Genes', y='Number of Detected DEGs', title='Detected DEG Count per gRNA Library')\n",
    "                           \n",
    "deg_count_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "embedded-spanish",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=30, repr.plot.height=10)\n",
    "merged_deg_plot | yuza_deg_plot | brunello_deg_plot | edwards_deg_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "intelligent-research",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_deg_count$deg_count %>% sum"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "banned-island",
   "metadata": {},
   "source": [
    "### Yuza Library Differential Gene Expression Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "mediterranean-meaning",
   "metadata": {},
   "outputs": [],
   "source": [
    "bar_plot_order = reorder(merged_deg_count$genes, merged_deg_count$deg_count)\n",
    "\n",
    "yuza_df <- yuza_df %>% filter(target_gene %in% merged_deg_count$genes)\n",
    "merged_df <- merged_df %>% filter(target_gene %in% merged_deg_count$genes)\n",
    "edwards_df <- edwards_df %>% filter(target_gene %in% merged_deg_count$genes)\n",
    "brunello_df <- brunello_df %>% filter(target_gene %in% merged_deg_count$genes)\n",
    "\n",
    "yuza_df$target_gene <- factor(yuza_df$target_gene, levels=merged_deg_count$genes)\n",
    "merged_df$target_gene <- factor(merged_df$target_gene, levels=merged_deg_count$genes)\n",
    "edwards_df$target_gene <- factor(edwards_df$target_gene, levels=merged_deg_count$genes)\n",
    "brunello_df$target_gene <- factor(brunello_df$target_gene, levels=merged_deg_count$genes)\n",
    "\n",
    "yuza_df <- yuza_df %>% arrange(target_gene)\n",
    "merged_df <- merged_df %>% arrange(target_gene)\n",
    "edwards_df <- edwards_df %>% arrange(target_gene)\n",
    "brunello_df <- brunello_df %>% arrange(target_gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "plastic-disaster",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# some manual filtering to improve the heatmap appearance\n",
    "yuza_df$logfc[yuza_df$logfc > 1] = 1\n",
    "yuza_df$logfc[yuza_df$logfc < -1] = -1\n",
    "\n",
    "yuza_df$log_padj[yuza_df$log_padj > 3] = 3\n",
    "\n",
    "# creating matrix for heatmaps\n",
    "logfc_mat = heatmap_matrix(yuza_df, uni_degs, 'logfc')\n",
    "padj_mat = heatmap_matrix(yuza_df, uni_degs, 'log_padj')\n",
    "\n",
    "# logfc_mat = logfc_mat[genes_levels, ]\n",
    "# padj_mat = padj_mat[genes_levels, ]\n",
    "# plotting log folc change and adjusted p-value\n",
    "yuza_heat = pheatmap(logfc_mat, color=logfc_palette, cluster_rows=FALSE, show_colnames=FALSE)\n",
    "pheatmap(padj_mat, color=padj_palette, cluster_rows=FALSE, show_colnames=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "assured-ballet",
   "metadata": {},
   "source": [
    "### Brunello Library Differential Gene Expression Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "premium-animation",
   "metadata": {},
   "outputs": [],
   "source": [
    "# some manual filtering to improve the heatmap appearance\n",
    "brunello_df$logfc[brunello_df$logfc > 1] = 1\n",
    "brunello_df$logfc[brunello_df$logfc < -1] = -1\n",
    "\n",
    "brunello_df$log_padj[brunello_df$log_padj > 3] = 3\n",
    "\n",
    "# creating matrix for heatmaps\n",
    "logfc_mat = heatmap_matrix(brunello_df, uni_degs, 'logfc')\n",
    "padj_mat = heatmap_matrix(brunello_df, uni_degs, 'log_padj')\n",
    "\n",
    "# plotting log folc change and adjusted p-value\n",
    "brunello_heat = pheatmap(logfc_mat, color=logfc_palette, cluster_rows=FALSE, show_colnames=FALSE)\n",
    "pheatmap(padj_mat, color=padj_palette, cluster_rows=FALSE, show_colnames=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "mental-english",
   "metadata": {},
   "source": [
    "### Edwards Library Differential Gene Expression Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "lovely-antibody",
   "metadata": {},
   "outputs": [],
   "source": [
    "# some manual filtering to improve the heatmap appearance\n",
    "edwards_df$logfc[edwards_df$logfc > 1] = 1\n",
    "edwards_df$logfc[edwards_df$logfc < -1] = -1\n",
    "\n",
    "edwards_df$log_padj[edwards_df$log_padj > 3] = 3\n",
    "\n",
    "# creating matrix for heatmaps\n",
    "logfc_mat = heatmap_matrix(edwards_df, uni_degs, 'logfc')\n",
    "padj_mat = heatmap_matrix(edwards_df, uni_degs, 'log_padj')\n",
    "\n",
    "# plotting log folc change and adjusted p-value\n",
    "edwards_heat = pheatmap(logfc_mat, color=logfc_palette, cluster_rows=FALSE, show_colnames=FALSE)\n",
    "pheatmap(padj_mat, color=padj_palette, cluster_rows=FALSE, show_colnames=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "empty-acoustic",
   "metadata": {},
   "source": [
    "## Venn Diagram of Differentially Expressed Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "driving-abuse",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "yuza_degs = find_deg_of_target_genes(yuza_res, expressed_genes)\n",
    "edwards_degs = find_deg_of_target_genes(edwards_res, expressed_genes)\n",
    "brunello_degs = find_deg_of_target_genes(brunello_res, expressed_genes)\n",
    "\n",
    "degs_list = list(Yuza=yuza_degs,\n",
    "                 Edwards=edwards_degs,\n",
    "                 Brunello=brunello_degs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "elegant-measurement",
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_venn = ggvenn(degs_list, c(\"Yuza\", \"Edwards\", \"Brunello\")) \n",
    "no_perc_venn = ggvenn(degs_list, show_percentage=FALSE, c(\"Yuza\", \"Edwards\", \"Brunello\")) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "operating-favor",
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_venn | no_perc_venn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "thousand-prize",
   "metadata": {},
   "outputs": [],
   "source": [
    "ggsave(\"normal_venn_diagram.eps\", plot=norm_venn, height=8, width=8, units=\"in\",device = \"eps\")\n",
    "ggsave(\"percentage_venn_diagram.eps\", plot=no_perc_venn, height=8, width=8, units=\"in\",device = \"eps\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "prerequisite-wright",
   "metadata": {},
   "source": [
    "# Differential Expression Analysis using all Cells Regardless of Library\n",
    "\n",
    "This can be viewed as an optimial performance profile as it contains 3x the cells than an individual library does on average."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "disabled-victory",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# some manual filtering to improve the heatmap appearance\n",
    "merged_df$logfc[merged_df$logfc > 1] = 1\n",
    "merged_df$logfc[merged_df$logfc < -1] = -1\n",
    "merged_df$log_padj[merged_df$log_padj > 3] = 3\n",
    "merged_df <- merged_df[!duplicated(merged_df), ]\n",
    "\n",
    "# creating matrix for heatmaps\n",
    "logfc_mat = heatmap_matrix(merged_df, merged_degs, 'logfc')\n",
    "padj_mat = heatmap_matrix(merged_df, merged_degs, 'log_padj')\n",
    "\n",
    "# creating annotation column dataframe\n",
    "anno_col_df = data.frame(Category = gene_metadata[rownames(logfc_mat), c(\"Category\")])\n",
    "rownames(anno_col_df) = rownames(logfc_mat)\n",
    "\n",
    "# plotting log folc change and adjusted p-value\n",
    "merged_heat = pheatmap(logfc_mat, color=logfc_palette, cluster_rows=FALSE, show_colnames=FALSE)\n",
    "pheatmap(padj_mat, color=padj_palette, cluster_rows=FALSE, show_colnames=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ordinary-service",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ggsave(\"yuza_deg_count.eps\", plot=yuza_deg_plot, height=8, width=8, units=\"in\",device = \"eps\")\n",
    "# ggsave(\"brunello_deg_count.eps\", plot=brunello_deg_plot, height=8, width=8, units=\"in\",device = \"eps\")\n",
    "# ggsave(\"edwards_deg_count.eps\", plot=edwards_deg_plot, height=8, width=8, units=\"in\",device = \"eps\")\n",
    "# ggsave(\"merged_deg_count.eps\", plot=merged_deg_plot, height=8, width=8, units=\"in\",device = \"eps\")\n",
    "# ggsave(\"deg_count.eps\", plot=deg_count_plot, height=8, width=8, units=\"in\", device = \"eps\")\n",
    "\n",
    "# ggsave(\"yuza_heatmaps.png\", plot=yuza_heat, height=8, width=8, units=\"in\",device = \"png\")\n",
    "# ggsave(\"brunello_heatmaps.png\", plot=brunello_heat, height=8, width=8, units=\"in\",device = \"png\")\n",
    "# ggsave(\"edwards_heatmaps.png\", plot=edwards_heat, height=8, width=8, units=\"in\",device = \"png\")\n",
    "# ggsave(\"merged_heatmaps.png\", plot=merged_heat, height=8, width=8, units=\"in\",device = \"png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "moved-municipality",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "791d1ca3-ca11-456f-8914-b7fe5da5fd12",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95bbe784-17b9-4670-837e-f873a0a55dea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "name,tags,-all",
   "main_language": "R",
   "notebook_metadata_filter": "-all"
  },
  "kernelspec": {
   "display_name": "edwards",
   "language": "R",
   "name": "edwards"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
